File: Problems/448. Find All Numbers Disappeared in an Array/solution.py
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums)
        
        if n == 1:
            return [nums]
        
        map = {}
        for num in nums: 
            map[num] = True
            
        result = []
        for i in range(n):
            if n - i not in map: 
                result.append(n - i)
                
        return result
        
        
# Time Complexity: O(n)
# Space Complexity: O(n)
# Time used: about 20 or 25 minutes. Not sure, solved it on a train lol
# Pattern: Arrays & Hashing

File: Problems/448. Find All Numbers Disappeared in an Array/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
// NOTES:
// - does the order of the elements in the output array matter? No.
// - Can I get an empty array? No, size will be at least 1. 
// - Can I assume that the numbers are only integers? Yes.

const findDisappearedNumbers = nums => {
  const seen = new Set();
  
  for (const num of nums) {
      seen.add(num);
  }
  
  const ans = [];
  for (let i = 1; i <= nums.length; i++) {
      if (!(seen.has(i))) {
          ans.push(i);
      }
  }
  
  return ans;
};

// Time complexity: O(n)
// Space complexity: O(n)
// Pattern: Arrays & Hashing
// TIme used: 12min